framework/include/csg/CSGBase.h

```diff
@@ -0,0 +1,417 @@
+
+  /**
+   * @brief Remove a cell from an existing universe
+   *
+   * @param universe universe from which to remove the cell
+   * @param cell cell to remove
+   */
+  void removeCellFromUniverse(const CSGUniverse & universe, const CSGCell & cell);
+
+  /**
+   * @brief Remove a list of cells from an existing universe
+   *
+   * @param universe universe from which to remove the cells
+   * @param cells list of references to cells to remove
+   */
+  void removeCellsFromUniverse(const CSGUniverse & universe,
+                               std::vector<std::reference_wrapper<const CSGCell>> & cells);
+
+  /**
+   * @brief Get all universe objects
+   *
+   * @return list of references to CSGUniverse objects in this CSGBase instance
+   */
+  std::vector<std::reference_wrapper<const CSGUniverse>> getAllUniverses() const
+  {
+    return _universe_list.getAllUniverses();
+  }
+
+  /**
+   * @brief Get a universe object by name
+   *
+   * @param name universe name
+   * @return reference to CSGUniverse object
+   */
+  const CSGUniverse & getUniverseByName(const std::string & name)
+  {
+    return _universe_list.getUniverse(name);
+  }
+
+  /**
+   * @brief Join another CSGBase object to this one. The cells of the root universe
+   * of the incoming CSGBase will be added to the existing root universe of this
+   * CSGBase.
+   *
+   * @param base pointer to a different CSGBase object
+   */
+  void joinOtherBase(std::unique_ptr<CSGBase> & base);
+
+  /**
+   * @brief Join another CSGBase object to this one. For the incoming CSGBase object,
+   * the root universe is added to this CSGBase object as a new non-root universe with
+   * the specified new name.
+   * Note: this newly created universe will not be connected to the root universe
+   * of this CSGBase object by default.
+   *
+   * @param base pointer to a different CSGBase object
+   * @param new_root_name_join new name for the universe generated from the incoming root universe
+   */
+  void joinOtherBase(std::unique_ptr<CSGBase> & base, std::string & new_root_name_join);
+
+  /**
+   * @brief Join another CSGBase object to this one. The root universe for the incoming CSGBase
+   * object is added to this CSGBase object as a non-root universe with a new name. The root
+   * universe of this CSGBase object will be renamed and designated as non-root.
+   * Note: upon completion of this join method, the root universe of this CSGBase
+   * object will be empty. Neither of the new non-root universes will be connected to the
+   * new root universe by default.
+   *
+   * @param base pointer to a different CSGBase object
+   * @param new_root_name_base new name for universe generated from this root universe
+   * @param new_root_name_join new name for the universe generated from the incoming root universe
+   */
+  void joinOtherBase(std::unique_ptr<CSGBase> & base,
+                     std::string & new_root_name_base,
+                     std::string & new_root_name_join);
```

Comment:
For all of the above here - are you taking ownership of `base`? If so, the unique_ptr should be passed in as the actual value (not a reference). If not, it should just be passed in as a plain reference (`CSGBase &`)

Reply:
Yes, taking ownership of base. Will update this to pass by value instead of by reference